compiler/src/main.rs

```diff
@@ -0,0 +1,208 @@
+use std::env;
+use std::ffi::{CStr, CString, OsStr};
+use std::fs::{self, File};
+use std::io::{self, Write};
+use std::mem::MaybeUninit;
+use std::path::{Path, PathBuf};
+use std::process::ExitCode;
+use std::str::{self, FromStr};
+
+use bpaf::Bpaf;
+
+mod error;
+use error::CompilerError;
+
+#[derive(Debug, Clone, Bpaf)]
+#[bpaf(options, version)]
+pub struct Options {
+    #[bpaf(external(output_type), optional)]
+    output_type: Option<OutputType>,
+
+    #[bpaf(long("linked"))]
+    /// Produce linked LLVM IR (in combination with --ll)
+    linked: bool,
+
+    #[bpaf(short('o'), argument("file"))]
+    /// Output path
+    output_path: Option<PathBuf>,
+
+    #[bpaf(positional("filename"))]
+    /// PTX file
+    ptx_path: String,
+}
+
+fn main() -> ExitCode {
+    main_core().map_or_else(
+        |e| {
+            eprintln!("Error: {}", e);
+            ExitCode::FAILURE
+        },
+        |_| ExitCode::SUCCESS,
+    )
+}
+
+fn main_core() -> Result<(), CompilerError> {
+    let opts = options().run();
+
+    let output_type = opts.output_type.unwrap_or_default();
+
+    if opts.linked && output_type != OutputType::LlvmIr {
+        println!("Warning: option --linked only makes sense when combined with --ll. Ignoring.");
+    }
+
+    let ptx_path = Path::new(&opts.ptx_path).to_path_buf();
+    check_path(&ptx_path)?;
+
+    let output_path = match opts.output_path {
+        Some(value) => value,
+        None => get_output_path(&ptx_path, &output_type)?,
+    };
+    check_path(&output_path)?;
+
+    let ptx = fs::read(&ptx_path).map_err(CompilerError::from)?;
+    let ptx = str::from_utf8(&ptx).map_err(CompilerError::from)?;
+    let llvm = ptx_to_llvm(ptx).map_err(CompilerError::from)?;
+
+    let output = match output_type {
+        OutputType::LlvmIr => {
+            if opts.linked {
+                get_linked_bitcode(&llvm)?
+            } else {
+                llvm.llvm_ir
+            }
+        }
+        OutputType::Elf => get_elf(&llvm)?,
+        OutputType::Assembly => get_assembly(&llvm)?,
+    };
+
+    write_to_file(&output, &output_path).map_err(CompilerError::from)?;
+    Ok(())
+}
+
+fn ptx_to_llvm(ptx: &str) -> Result<LLVMArtifacts, CompilerError> {
+    let ast = ptx_parser::parse_module_checked(ptx).map_err(CompilerError::from)?;
+    let module = ptx::to_llvm_module(ast).map_err(CompilerError::from)?;
+    let bitcode = module.llvm_ir.write_bitcode_to_memory().to_vec();
+    let linked_bitcode = module.linked_bitcode().to_vec();
+    let llvm_ir = module.llvm_ir.print_module_to_string().to_bytes().to_vec();
+    Ok(LLVMArtifacts {
+        bitcode,
+        linked_bitcode,
+        llvm_ir,
+    })
+}
+
+#[derive(Debug)]
+struct LLVMArtifacts {
+    bitcode: Vec<u8>,
+    linked_bitcode: Vec<u8>,
+    llvm_ir: Vec<u8>,
+}
+
+fn get_arch() -> Result<CString, CompilerError> {
```

Comment:
It's absolutely a valid flow to compile without AMD GPU or compile for a GPU you don't have installed.
This is a good default, but we need some more:
* An option to pass the architecture name to compile for
* If any of the hip runtime functions to detect gpu fails (or there is no gpu) then use a default: gfx1100